### PR TITLE
[4706] Add date separators for empty threads

### LINKED ISSUE #4706: Add thread separators for empty threads in `V6`
### Context & proposal

We have implemented the feature in `v5` already, however, it needs to be replicated in `v6` as well.

PR implementing the V5 version seen here: https://github.com/GetStream/stream-chat-android/pull/4629

### Acceptance criteria

To accept this task we need to complete the following items:

- [ ] Implement the given feature.

### PR DESCRIPTION
**Pull request by MarinTolic**
### 🎯 Goal

Closes #4706

### 🛠 Implementation details

Added a new state item for empty thread placeholders, a new boolean `showThreadSeparatorInEmptyThread` used for displaying thread separators in empty threads and the necessary logic inside `MessageListController.groupMessages()`

### 🎨 UI Changes

Compose:

| Before | After |
| --- | --- |
| Screenshot_1679326316 | Screenshot_1679326144 | 

UI Comp:

| Before | After |
| --- | --- |
| Screenshot_1679326369 | Screenshot_1679326131 | 

#### Test A - Check that no bugs were introduced

1. Open an empty thread, send a response
2. Open a thread that already has messages, send a response
3. Repeat for the other SDK

Expectation: No anomalous behavior should occur, eveything should work as it has previously.

#### Test B - Check the new funtionality

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt	(revision 3080a18c7d8ef880e3b434c66d3d13fa218e30dc)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt	(date 1679324795455)
@@ -22,12 +22,14 @@
 import androidx.compose.foundation.layout.fillMaxWidth
 import androidx.compose.foundation.layout.padding
 import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material.Card
 import androidx.compose.material.Surface
 import androidx.compose.material.Text
 import androidx.compose.runtime.Composable
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.graphics.Brush
+import androidx.compose.ui.graphics.Shape
 import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.text.style.TextAlign
 import androidx.compose.ui.unit.dp
@@ -93,7 +95,15 @@
         )
     },
     typingIndicatorContent: @Composable (TypingItemState) -> Unit = { },
-    emptyThreadPlaceholderItemContent: @Composable (EmptyThreadPlaceholderItemState) -> Unit = { },
+    emptyThreadPlaceholderItemContent: @Composable (EmptyThreadPlaceholderItemState) -> Unit = {
+        Card(
+            modifier = Modifier.padding(8.dp),
+            backgroundColor = ChatTheme.colors.primaryAccent,
+            shape = RoundedCornerShape(4.dp)
+        ) {
+            Text(text = "Hi, I am a placeholder")
+        }
+    },
 ) {
     when (messageListItemState) {
         is DateSeparatorItemState -> dateSeparatorContent(messageListItemState)
Index: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/feature/messages/list/adapter/MessageListItemViewHolderFactory.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/feature/messages/list/adapter/MessageListItemViewHolderFactory.kt b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/feature/messages/list/adapter/MessageListItemViewHolderFactory.kt
--- a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/feature/messages/list/adapter/MessageListItemViewHolderFactory.kt	(revision 3080a18c7d8ef880e3b434c66d3d13fa218e30dc)
+++ b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/feature/messages/list/adapter/MessageListItemViewHolderFactory.kt	(date 1679324587487)
@@ -19,6 +19,7 @@
 import android.view.View
 import android.view.ViewGroup
 import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.databinding.StreamUiCustomTestBinding
 import io.getstream.chat.android.ui.feature.messages.list.GiphyViewHolderStyle
 import io.getstream.chat.android.ui.feature.messages.list.MessageListItemStyle
 import io.getstream.chat.android.ui.feature.messages.list.MessageReplyStyle
@@ -37,6 +38,7 @@
 import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemViewType.THREAD_PLACEHOLDER
 import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemViewType.THREAD_SEPARATOR
 import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemViewType.TYPING_INDICATOR
+import io.getstream.chat.android.ui.feature.messages.list.adapter.internal.DecoratedBaseMessageItemViewHolder
 import io.getstream.chat.android.ui.feature.messages.list.adapter.internal.MessageListItemViewTypeMapper
 import io.getstream.chat.android.ui.feature.messages.list.adapter.viewholder.attachment.AttachmentFactoryManager
 import io.getstream.chat.android.ui.feature.messages.list.adapter.viewholder.decorator.internal.DecoratorProvider
@@ -54,6 +56,7 @@
 import io.getstream.chat.android.ui.feature.messages.list.adapter.viewholder.internal.SystemMessageViewHolder
 import io.getstream.chat.android.ui.feature.messages.list.adapter.viewholder.internal.ThreadSeparatorViewHolder
 import io.getstream.chat.android.ui.helper.transformer.ChatMessageTextTransformer
+import io.getstream.chat.android.ui.utils.extensions.streamThemeInflater
 
 /**
  * A factory class designed to create ViewHolders for the RecyclerView
@@ -163,7 +166,7 @@
             GIPHY -> createGiphyMessageItemViewHolder(parentView)
             SYSTEM_MESSAGE -> createSystemMessageItemViewHolder(parentView)
             ERROR_MESSAGE -> createErrorMessageItemViewHolder(parentView)
-            THREAD_PLACEHOLDER -> createEmptyMessageItemViewHolder(parentView)
+            THREAD_PLACEHOLDER -> createCustomView(parentView)
             LINK_ATTACHMENTS -> createLinkAttachmentsViewHolder(parentView)
             GIPHY_ATTACHMENT -> createGiphyAttachmentViewHolder(parentView)
             FILE_ATTACHMENTS -> createFileAttachmentsViewHolder(parentView)
@@ -172,6 +175,20 @@
         }
     }
 
+    private fun createCustomView(parentView: ViewGroup): BaseMessageItemViewHolder<out MessageListItem>  {
+        val binding = StreamUiCustomTestBinding.inflate(
+            parentView.streamThemeInflater,
+            parentView,
+            false
+        )
+
+        return object : DecoratedBaseMessageItemViewHolder<MessageListItem.ThreadPlaceholderItem>(binding.root, decoratorProvider.decorators) {
+            override fun bindData(data: MessageListItem.ThreadPlaceholderItem, diff: MessageListItemPayloadDiff?) {
+                super.bindData(data, diff)
+            }
+        }
+    }
+
     /**
      * Creates the custom attachments view holder.
      *
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(revision 3080a18c7d8ef880e3b434c66d3d13fa218e30dc)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(date 1679325897233)
@@ -57,7 +57,8 @@
         MessageListViewModelFactory(
             cid = args.cid,
             messageId = args.messageId,
-            parentMessageId = args.parentMessageId
+            parentMessageId = args.parentMessageId,
+            showDateSeparatorInEmptyThread = true,
         )
     }
     private val chatViewModelFactory: ChatViewModelFactory by lazy { ChatViewModelFactory(args.cid) }
Index: stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml b/stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml
new file mode 100644
--- /dev/null	(date 1679325950876)
+++ b/stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml	(date 1679325950876)
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+    Copyright (c) 2014-2023 Stream.io Inc. All rights reserved.
+
+    Licensed under the Stream License;
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+-->
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:padding="12dp"
+    >
+
+    <androidx.cardview.widget.CardView
+        android:paddingVertical="12dp"
+        android:paddingHorizontal="8dp"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        app:cardBackgroundColor="@color/stream_ui_accent_blue"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintHorizontal_bias="0"
+        >
+
+        <TextView
+            android:layout_gravity="start"
+            android:padding="8dp"
+            android:layout_width="wrap_content"
+            android:id="@+id/customText"
+            android:layout_height="wrap_content"
+            android:gravity="bottom"
+            android:text="Custom placeholder item test"
+            />
+
+    </androidx.cardview.widget.CardView>
+
+
+</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
Index: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
--- a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(revision 3080a18c7d8ef880e3b434c66d3d13fa218e30dc)
+++ b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(date 1679325897237)
@@ -80,7 +80,8 @@
             channelId = requireNotNull(intent.getStringExtra(KEY_CHANNEL_ID)),
             deletedMessageVisibility = DeletedMessageVisibility.ALWAYS_VISIBLE,
             messageId = intent.getStringExtra(KEY_MESSAGE_ID),
-            parentMessageId = intent.getStringExtra(KEY_PARENT_MESSAGE_ID)
+            parentMessageId = intent.getStringExtra(KEY_PARENT_MESSAGE_ID),
+            showDateSeparatorInEmptyThread = true,
         )
     }
 
```

</details>

1. Apply the patch above
2. Open a new thread, post a new message
3. Open an existing thread with messages, post a new message
4. Repeat for both apps

Expectation: Date separators and placeholders should be visible inside empty threads. Placeholder items should disappear once a message has been sent. They should not appear in threads with existing messages.

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

gif

### CODE REVIEW

**Review by skydoves — APPROVED** [review]
Looks good to me!

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/viewmodel/messages/MessageListViewModelFactory.kt [hunk 1]
```diff
@@ -48,6 +48,8 @@ import io.getstream.chat.android.ui.common.utils.AttachmentConstants
  * @param dateSeparatorHandler Handler that determines when the date separators should be visible.
  * @param threadDateSeparatorHandler Handler that determines when the thread date separators should be visible.
  * @param messagePositionHandler Determines the position of the message inside a group.
+ * @param showDateSeparatorInEmptyThread Configures if we show a thread separator when threads are empty or not.
+ * Adds the separator item when the value is `true`.
```
Comment (TGazica) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/viewmodel/messages/MessageListViewModelFactory.kt:52]:
```suggestion
 * @param showDateSeparatorInEmptyThread Determines if we show a thread separator when threads are empty or not if true, hides them if false.
```

Same

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt [hunk 2]
```diff
@@ -58,6 +60,8 @@ import io.getstream.chat.android.ui.common.state.messages.list.TypingItemState
  * @param systemMessageContent Composable that represents system messages.
  * @param messageItemContent Composable that represents regular messages.
  * @param typingIndicatorContent Composable that represents a typing indicator.
+ * @param emptyThreadPlaceholderItemContent Composable that represents placeholders inside of an empty thread.
+ * This content is disabled by default and can be enabled via [MessageListController.showDateSeparatorInEmptyThread].
```
Comment (TGazica) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt:64]: Maybe add `MessageListViewModel` or `MessageListViewModelFactory`. Users using our default components will probably create the VM through them. Users using `MessageListControler` will probably get that from it's docs.

**Review by TGazica — CHANGES_REQUESTED** [review]
Code looks good and works as expected, left a few comments.

File: stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessageListController.kt [hunk 3]
```diff
@@ -136,6 +137,8 @@ import io.getstream.chat.android.ui.common.state.messages.Flag as FlagMessage
  * @param dateSeparatorHandler Determines the visibility of date separators inside the message list.
  * @param threadDateSeparatorHandler Determines the visibility of date separators inside the thread.
  * @param messagePositionHandler Determines the position of the message inside a group.
+ * @param showDateSeparatorInEmptyThread Configures if we show a thread separator when threads are empty or not.
+ * Adds the separator item when the value is `true`.
```
[the remark this answers is not in the document]
Comment (MarinTolic) [reply stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessageListController.kt]: Updated

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt (hunk 2 above)
Comment (MarinTolic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageContainer.kt]: Updated

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessagesViewModelFactory.kt [hunk 4]
```diff
@@ -57,6 +57,8 @@ import io.getstream.chat.android.ui.common.utils.AttachmentConstants
  * @param dateSeparatorHandler Handler that determines when the date separators should be visible.
  * @param threadDateSeparatorHandler Handler that determines when the thread date separators should be visible.
  * @param messagePositionHandler Determines the position of the message inside a group.
+ * @param showDateSeparatorInEmptyThread Configures if we show a thread separator when threads are empty or not.
+ * Adds the separator item when the value is `true`.
```
[the remark this answers is not in the document]
Comment (MarinTolic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessagesViewModelFactory.kt]: updated

**Review by TGazica — APPROVED** [review]

**Pull request merged**